### PR TITLE
During IBD, prune as much as possible until we get close to where we will eventually keep blocks

### PR DESCRIPTION
**Pull request by luke-jr**
This should reduce pruning flushes even more, speeding up IBD with pruning on systems that have a sufficient dbcache.

Assumes 1 MB per block between tip and best header chain. Simply adds this to the buffer pruning is trying to leave available, which results in pruning almost everything up until we get close to where we need to be keeping blocks.

### CODE REVIEW

**Comment by jonatack** [comment]
Interesting, will have a look.

**Comment by jonasschnelli** [comment]
Nice!
Concept ACK.
Could we further reduce flushes by allowing a user parameter (`-pruneflushbuffer` [or similar])? Some users are probably okay with providing a few GBs of extra space for speeding up pruning (assume hight dbcache) (but wan't the extra space be freed after IBD).
Or we could automatically use he dbcache size as the buffer size.

**Comment by Sjors** [comment]
I also like the idea of temporally assigning more RAM for IBD. This could even be done by default for GUI users (with opt-out in the intro screen).

**Comment by luke-jr** [comment]
`dbcache` adjustments are ultimately an unrelated feature. I see #19873 as the next step in that area.

**Comment by practicalswift** [comment]
Concept ACK

Would be nice to see this benchmarked.

File: src/validation.cpp [hunk 1]
```diff
@@ -4018,10 +4018,12 @@ void BlockManager::FindFilesToPrune(std::set<int>& setFilesToPrune, uint64_t nPr
...
         // To avoid excessive prune events negating the benefit of high dbcache
         // values, we should not prune too rapidly.
-        // So when pruning in IBD, increase the buffer a bit to avoid a re-prune too soon.
-        if (is_ibd) {
-            // Since this is only relevant during IBD, we use a fixed 10%
-            nBuffer += nPruneTarget / 10;
+        // So when pruning in IBD, increase the buffer to avoid a re-prune too soon.
+        if (is_ibd && target_sync_height > (uint64_t)chain_tip_height) {
+            // Since this is only relevant during IBD, we assume blocks are at least 1 MB on average
+            static constexpr uint64_t average_block_size = 1000000;  /* 1 MB */
```
Comment (rebroad) [inline src/validation.cpp:4024]: why assume when you can calculate?

Comment (luke-jr) [reply src/validation.cpp:4024]: AFAIK we don't have access to future block sizes at this point.

**Comment by stickies-v** [comment]
Without rebase, compiling fails for me:

```shell
$ make clean && ./configure CXXFLAGS="-O0 -g" CFLAGS="-O0 -g" && make -j 9
...
touch src/config/bitcoin-config.h.in
  CXX      util/libbitcoin_util_a-moneystr.o
  CXX      util/libbitcoin_util_a-readwritefile.o
  CXX      util/libbitcoin_util_a-settings.o
  CXX      util/libbitcoin_util_a-serfloat.o
  CXX      util/libbitcoin_util_a-spanparsing.o
  CXX      util/libbitcoin_util_a-strencodings.o
  CXX      util/libbitcoin_util_a-string.o
  CXX      util/libbitcoin_util_a-url.o
make[3]: *** No rule to make target `libunivalue.la'.  Stop.
make[2]: *** [univalue/libunivalue.la] Error 2
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [all-recursive] Error 1
make: *** [all-recursive] Error 1
```

After `git rebase master`, everything runs smoothly again. Will report back shortly with my review results!

**Comment by MarcoFalke** [comment]
@stickies-v If you want to compile older commits of master, you'll need to `make distclean` first.

**Comment by luke-jr** [comment]
> Without rebase, compiling fails for me:

Note: It's usually best to merge PRs into master for testing. Sometimes (bugfixes) they can be based on very old commits (when the bug was introduced); and if there's a silent problem with the merge, you'd want to notice that too.

>@stickies-v If you want to compile older commits of master, you'll need to make distclean first.

Can we fix this build system bug?

**Comment by theStack** [comment]
Concept ACK

**Comment by 0xB10C** [comment]
Concept ACK

We'll be doing a Bitcoin Core PR Review club covering this PR on the 29th: [bitcoincore.reviews/20827](https://bitcoincore.reviews/20827)

**Comment by 0xB10C** [comment]
https://github.com/bitcoin/bitcoin/pull/12404 (not merged) may also be interesting to reviewers. @Sjors did run a few benchmarks back then and found https://github.com/bitcoin/bitcoin/pull/11658 (which this PR overrides) to be the better option at that time.

I'm planing on benchmarking this change too.

**Comment by fjahr** [comment]
Concept ACK, I read the PR Review Club and will review after rebase.

**Comment by 0xB10C** [comment]
With a prune just before IBD, there is more blk and undo data available when running master than with this change. One potential issue that was mentioned during the PR review club:

On non-mainnet networks, the 1MB per block assumption doesn't hold. When we prune with a large `nBuffer` just before IBD ends, the prune might cause problems on larger reorgs as not enough undo data is present to revert back to the last shared block. Testnet has seen large reorgs in the past. 

Might be worth thinking about potential alternatives to increasing the `nBuffer` by `1MB * remaining_blocks`. There might be cleaner ways to implement "prune _everything_ as long as we are not close to IBD being done". Haven't come up with a practical threshold for "IBD being close to done" yet that works for all networks.

**Comment by luke-jr** [comment]
The 1 MB per block assumption would mean at least 550 blocks. Testnet has deeper reorgs than that?

**Comment by 0xB10C** [comment]
> The 1 MB per block assumption would mean at least 550 blocks. Testnet has deeper reorgs than that?

I haven't observed this personally, but yes testnet has been unreliable in regards to deep reorgs in the past. That's one motivation for SigNet in BIP 325. Also:
- https://github.com/lightningnetwork/lnd/issues/3060#issuecomment-491096343
- https://blog.lopp.net/the-block-storms-of-bitcoins-testnet/

**Comment by 0xB10C** [comment]
> benchmarking [0xB10C@db2a9e7](https://github.com/0xB10C/bitcoin/commit/db2a9e71a748c4806b85f767441fb3f090d62163) (PR) vs [0xB10C@2e01b69](https://github.com/0xB10C/bitcoin/commit/2e01b6986099715afa40ed6464da4b321b630e9c) (mergebase; MB).
>
> I have 8 different prune and dbcache configurations set-up. Each configuration runs three times per binary (PR and MB). That's 8 configs * 3 runs * 2 binaries = 48. I expect this to take about 6 days. Configurations:
>
> ```
> 1. -dbcache=300 -prune=550
> 2. -dbcache=4000 -prune=550
>
> 3. -dbcache=300 -prune=1100
> 4. -dbcache=4000 -prune=1100
>
> 5. -dbcache=300 -prune=2200
> 6. -dbcache=4000 -prune=2200
>
> 7. -dbcache=300 -prune=4400
> 8. -dbcache=4000 -prune=4400
> ```
>
> Benchmarks run on a dedicated, properly cooled, and otherwise idle machine with two HDDs in a RAID0. The node syncs from a local node on the same machine and disks (connected via `addnode` and otherwise `connect=0`). The sync starts at block height 500_000 and ends at height 600_000. Between each run, the machine idles 5 minutes. The debug.log, with extra `debug=coindb debug=prune` logging, of each run is saved. I'll post results and graphs next week.



**tl;dr**: With some configurations, an IBD performance improvement of 15% or more between block 500k and 600k was measured.


### Number of prunes, pruned files, dbcache flushes

This PR makes pruning more aggressive. We prune more blk/rev files in a single prune operation to minimize the total number of prune operations needed. This also reduces the number of expensive dbcache flushes as each prune operation requires a dbcache flush.


| config                     |   MB prunes |   PR prunes | prunes   |   MB pruned files |   PR pruned files | pruned files   |   MB flushes |   PR flushes | flushes   |
|----------------------------|-------------|-------------|----------|-------------------|-------------------|----------------|--------------|--------------|-----------|
| 0: prune=550 dbcache=300   |         726 |         670 | 92.29%   |                 1 |              1.08 | 1.08x          |          726 |          670 | 92%       |
| 1: prune=550 dbcache=4000  |         726 |         670 | 92.29%   |                 1 |              1.08 | 1.08x          |          726 |          670 | 92%       |
| 2: prune=1100 dbcache=300  |         723 |         150 | 20.75%   |                 1 |              4.83 | 4.83x          |          723 |          150 | 21%       |
| 3: prune=1100 dbcache=4000 |         723 |         150 | 20.75%   |                 1 |              4.83 | 4.83x          |          723 |          150 | 21%       |
| 4: prune=2200 dbcache=300  |         358 |          58 | 16.2%    |                 2 |             12.47 | 6.23x          |          359 |           76 | 21%       |
| 5: prune=2200 dbcache=4000 |         358 |          58 | 16.2%    |                 2 |             12.47 | 6.23x          |          358 |           58 | 16%       |
| 6: prune=4400 dbcache=300  |         175 |          26 | 14.86%   |                 4 |             27.69 | 6.92x          |          178 |           69 | 39%       |
| 7: prune=4400 dbcache=4000 |         175 |          26 | 14.86%   |                 4 |             27.69 | 6.92x          |          175 |           26 | 15%       |

In the IBDs for **configurations 0 & 1** (prune = 550MB; minimum prune target), we only prune and flush about 8% less in _PR_ than in _MB_. With _MB_, every prune operation prunes 1 blk/rev file pairs. _PR_ prune operations sometimes prune 2 blk/rev files, averaging at 1.08 files per prune. Increasing the dbcache doesn't affect the number of flushes. Likely because the max cache size is not reached.

Similarly, a larger dbcache doesn't affect **configurations 2 & 3**. However, _PR_ only has 21% of the flushes and prune operations of _MB_. _MB_ prunes 1 file pair per prune operation while _PR_ prunes 4.83 file pairs on average.

We first see a larger dbcache affecting the number of flushes in **configurations 4 & 5**. Configuration 4 (c4) has a dbcache of 300MB and configuration 5 (c5) a dbcache of 4000MB. With c4, the number of flushes of _PR_ compared to _MB_ is only 21%. With c5, it's 16%. _PR_ prunes on average 12.47 file pairs per prune operation, while _MB_ prunes 2 pairs.

With **configurations 6 & 7** _PR_ prunes 27.7 file pairs on average while _MB_ prunes 4 pairs. However, in configuration 6 (c6) the small dbcache (300MB) is a limiting factor. _PR_ flushes 69 times with c6 and 26 times with configuration 7 (dbcache=4000MB) compared to 178 (c6) and 175 (c7) flushes with _MB_.


### IBD performance (between block 500000 and 600000)

Fewer dbcache flushes due to fewer prune operations should yield IBD performance improvements.

| Configuration              | MB run 1   | MB run 2   | MB run 3   | MB mean   | PR run 1   | PR run 2   | PR run 3   | PR mean   | improvement   |
|----------------------------|------------|------------|------------|-----------|------------|------------|------------|-----------|---------------|
| 0: prune=550 dbcache=300   | 9799s      | 10046s     | 10003s     | 9949s     | 9478s      | 9531s      | 9518s      | 9509s     | 4.4%          |
| 1: prune=550 dbcache=4000  | 10219s     | 10047s     | 9915s      | 10060s    | 9649s      | 9549s      | 9712s      | 9636s     | 4.2%          |
| 2: prune=1100 dbcache=300  | 9970s      | 10191s     | 10073s     | 10078s    | 8249s      | 8467s      | 8310s      | 8342s     | 17.2%         |
| 3: prune=1100 dbcache=4000 | 10084s     | 9887s      | 9976s      | 9982s     | 8652s      | 8483s      | 8329s      | 8488s     | 15.0%         |
| 4: prune=2200 dbcache=300  | 8991s      | 9071s      | 8817s      | 8959s     | 7494s      | 7687s      | 7496s      | 7559s     | 15.6%         |
| 5: prune=2200 dbcache=4000 | 8973s      | 8865s      | 9063s      | 8967s     | 7405s      | 7169s      | 7661s      | 7411s     | 17.4%         |
| 6: prune=4400 dbcache=300  | 8378s      | 8102s      | 7972s      | 8150s     | 7645s      | 7533s      | 7515s      | 7564s     | 7.2%          |
| 7: prune=4400 dbcache=4000 | 8144s      | 8263s      | 8002s      | 8136s     | 6833s      | 6982s      | 6792s      | 6869s     | 15.6%         |

We see these performance improvements for all configurations when comparing the IBD time from _PR_ to _MB_.

In configurations 0 (c0) and 1 (c1), the increased dbcache had no effect. We saw 8% fewer prunes and flushes with _PR_ than _MB_ for both configurations. We measured a 4.4% (c0) and a 4.2% (c1) improvement in IBD performance with _PR_.

image


In configurations 2 (c2) and 3 (c3), _PR_ had only 21% of the flush and prune operations than _MB_ had (see above). We measured a performance improvement of 17.2% for c2 and a 15% for c3. The difference of 2.2% stems likely from _PR run 1_ of c3, which took about 400s (~5%) longer than _PR run 1_ of c2.

image


The smaller dbcache of configuration 4 (c4) was noticeable in the number of flushes and prunes already. The _PR_ showed a 15.6% improvement over _MB_ in c4 and a 17.4% improvement in c5.

image


With configurations 6 (c6) and 7 (c7), the effect of the smaller dbcache of c6 became more noticeable. We saw a 7.2% improvement for _PR_ with c6 compared to a 15.6% improvement with c7.

image

### Conclusion

_PR_ with a larger prune target does reduce the number of flushes significantly compared to _MB_ by deleting more blk/rev file pairs per prune operation. However, the default dbcache (of 300MB) can fill up with larger prune targets. Here, increasing the dbcache is helpful. Given the right configuration, fewer flushes significantly impact the IBD performance between block 500000 and 600000. Often, a performance improvement of over 15% can be measured.

This PR is worth pursuing. 

---

I'm running full (genesis to 710k) IBD benchmarks now.

**Review by ghost — APPROVED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/20827/commits/24f3936337de3afb4fa56efc83009e2527d22df0

Review by @0xB10C is helpful although I was confused at some places if MB is for Master Branch or Mega Byte.

**Comment by 0xB10C** [comment]
> Review by @0xB10C is helpful although I was confused at some places if MB is for Master Branch or Mega Byte.

Hm, good point. MB actually stands for mergebase (the commit right before the changes from this PR) here.

**Comment by 0xB10C** [comment]
I ran full IBD benchmarks (block 0 to 710k compared to 500k to 600k in https://github.com/bitcoin/bitcoin/pull/20827#issuecomment-1017603590) with these three configurations:

```
1. -dbcache=300 -prune=550
2. -dbcache=4000 -prune=4400
3. -dbcache=4000 -prune=8800
```
I ran each configuration two times (compared to three times in https://github.com/bitcoin/bitcoin/pull/20827#issuecomment-1017603590) for both the PR binary and the MB binary. That's 3 configurations * 2 binaries * 2 runs = 12 full IBDs. The setup and hardware was the same as in the 500k-to-600k benchmark. However, I've started with an empty datadir compared to a datadir pre-synced to block 500k in the previous benchmark.


**tl;dr:** I measured a 6% full-IBD speedup with a default dbcache and the minimum pruning target and a 17.2% full-IBD speedup with a dbcache of 4000MB and a prune target of 8800MB.
 


Number of prunes, pruned files, dbcache flushes

**Comment by luke-jr** [comment]
Rebased

**Comment by achow101** [comment]
Are you still working on this?

**Comment by luke-jr** [comment]
Rebased

**Comment by kristapsk** [comment]
Concept ACK

**Comment by BenWestgate** [comment]
I have a project that syncs bitcoin core on an external drive and this would really help a lot of people testing it with smaller flash drives and higher RAM.  So much so that I've made scripts to do manual pruning (back to 550mib) only when disk space is getting low to emulate this.

File: src/node/blockstorage.cpp [hunk 2]
```diff
@@ -298,6 +298,7 @@ void BlockManager::FindFilesToPrune(
     // Distribute our -prune budget over all chainstates.
     const auto target = std::max(
         MIN_DISK_SPACE_FOR_BLOCK_FILES, GetPruneTarget() / chainman.GetAll().size());
+    const uint64_t target_sync_height = chainman.m_best_header->nHeight;
```
Comment (andrewtoth) [inline src/node/blockstorage.cpp:301]: nit: Any reason this declaration is up here and not directly above where it is used below?

File: src/node/blockstorage.cpp [hunk 3]
```diff
@@ -320,10 +321,13 @@ void BlockManager::FindFilesToPrune(
         // On a prune event, the chainstate DB is flushed.
         // To avoid excessive prune events negating the benefit of high dbcache
         // values, we should not prune too rapidly.
-        // So when pruning in IBD, increase the buffer a bit to avoid a re-prune too soon.
-        if (chainman.IsInitialBlockDownload()) {
-            // Since this is only relevant during IBD, we use a fixed 10%
-            nBuffer += target / 10;
+        // So when pruning in IBD, increase the buffer to avoid a re-prune too soon.
+        const auto chain_tip_height = chain.m_chain.Height();
```
Comment (andrewtoth) [inline src/node/blockstorage.cpp:325]: nit: could remove the cast in the next line with
```suggestion
        const uint64_t chain_tip_height = chain.m_chain.Height();
```

**Comment by andrewtoth** [comment]
ACK d298ff8b62b2624ed390c8a2f905c888ffc956ff

Ran benchmark with hyperfine:
```
hyperfine --show-output --parameter-list commit 96ec3b67a7a7f968d002e13d6fc227f69b7f07d7,d298ff8b62b2624ed390c8a2f905c888ffc956ff --setup 'git checkout {commit} && make -j$(nproc) src/bitcoind' --prepare 'sync; sudo /sbin/sysctl vm.drop_caches=3; rm -r /home/user/.bitcoin/blocks /home/user/.bitcoin/chainstate' -M 3 './src/bitcoind -dbcache=2000 -prune=2000 -printtoconsole=0 -stopatheight=800000'
```

Results were 21% faster on this branch :rocket: 
```
Summary
  ./src/bitcoind -dbcache=2000 -prune=2000 -printtoconsole=0 -stopatheight=800000 (commit = d298ff8b62b2624ed390c8a2f905c888ffc956ff) ran
    1.21 ± 0.00 times faster than ./src/bitcoind -dbcache=2000 -prune=2000 -printtoconsole=0 -stopatheight=800000 (commit = 96ec3b67a7a7f968d002e13d6fc227f69b7f07d7)
```

**Comment by fjahr** [comment]
utACK d298ff8b62b2624ed390c8a2f905c888ffc956ff

While there may be slightly better configurations possible, as @0xB10C pointed out above, I think the 1MB assumption is alright and this is a clear improvement already, so this can be merged and potentially improved later.

**Comment by achow101** [comment]
ACK d298ff8b62b2624ed390c8a2f905c888ffc956ff

**Comment by vostrnad** [comment]
15% improvement in IBD performance in some configurations seems significant enough for a release note, no?
